src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+// (extracted expression, dynamic/inner type) like ExpandResult
+// Htyp.t is like the ground type
+
+//type extract_result = option((ocaml_exp, HTyp.t));
+type t = (ocaml_exp, HTyp.t)
+and ocaml_exp = string;
+
+/* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+//TODO: replace == with HTyp.consistent
+let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): t =>
+  switch (de) {
+  | EmptyHole(_) => failwith("Exp: Empty Hole")
+  | NonEmptyHole(_) => failwith("Exp: Non-Empty Hole")
+  // used for variables that are keywords in Hazel, like "let" and "case"
+  | Keyword(_) => failwith("Exp: Incomplete Program (Keyword)")
+  // an expression with a free variable cannot be extracted to OCaml because free variables cause type errors in OCaml
+  | FreeVar(_, _, _, _) => failwith("Exp: FreeVar is not allowed")
+  | BoundVar(x) =>
```

Comment:
need to add a check that the variable is a valid ocaml variable name

Reply:
let's add the check to pattern, so whenever it's declared, it should be correct. I'll use Var.is_valid to check, because ocaml variable follows the same regular expression.

Reply:
Sorry I can't use Var.is_valid, because ocaml variable shouldn't start with A-Z